Front end working and other changes

vdavez opened the pull request:
This should be ready for review. Here's what's happening. In addition to the changes made in #14, here are the salient changes. 
1. Added a management command to allow for loading FY16 Q1 data.
2. Cleaned up the API.
3. Started splitting out the `Opportunity` model (by splitting out `OSBUAdvisor`).
4. Added the design standards library and mocked up the home page to start to resemble https://projects.invisionapp.com/share/KD530ALCM#/screens/118305787/comments
   <img width="1109" alt="screen shot 2015-11-27 at 12 43 29 pm" src="https://cloud.githubusercontent.com/assets/4153048/11446458/32f2fcca-9505-11e5-85a5-c1cff605cc98.png">
5. Reverted back to the full-length choice tuples in `Opportunity` because I wanted words not numbers/abbreviations in the API serialization
6. Made other changes to the `Opportunity` model, such as:
   - DecimalField for the price data
   - Changing FY data to integers
   - Adding sane defaults to required fields
   - Adding validator for NAICS code.
7. Added some fixtures (including a users.yaml fixture for an admin user).
8. Added a bulk publication for admins. 

Note, I had some trouble with the migrations, so I ended up wiping the db and the migrations folder, and restarting with migrations...

Commit:
first pass at wireframing

Commit:
first pass at wireframing

Commit:
added fixtures

Commit:
Beginning of load command and defaults

This commit adds the beginnings of a load command for opportunities from
Q1 FY16. It also starts updating the model to require more fields but
providing defaults.

Commit:
Changed admin header; split out OSBU Advisor model

Split out the OSBU Advisor model, which will likely extend the User
model in a future commit. Also changed the admin page header.

Commit:
Reverted tuples; added OSBU serialization

One of the downsides of the tuple simplification is that it's less
verbose when using the API. So, I reverted back to descriptive tuples
and included the OSBU_Advisor serialization for the API.

Commit:
move to DecimalField

Commit:
move to DecimalField

Commit:
some model cleanup

Commit:
updated admin interface

Commit:
fixed OSBU serialization

Commit:
fixed tests to account for model name change

Commit:
trying to fix travis

Commit:
successful loading of Q1 data

Commit:
JSON loaded on home

Commit:
getting closer

Commit:
ready

khandelwal commented on forecast-admin/forecast/opportunities/admin.py:
```diff
@@ -34,10 +39,8 @@ def get_queryset(self, request):
+
+@admin.register(OSBUAdvisor)
+class OSBUAdvisor(admin.ModelAdmin):
+    pass
```
I'd stick a comment in here to indicate why this is just 'pass'

vdavez replied:
Fixed per 93d410d

khandelwal commented on forecast-admin/forecast/opportunities/fixtures/create_offices.py:
```diff
@@ -0,0 +1,49 @@
+import json
```
I'd put a docstring at the very beginning of this script indicating when this script should be used, and how.

khandelwal replied:
One way to simply explain what this does would be to rename this file to: office_fixture_generator.py (which is what I assume this is doing).

vdavez replied:
Fixed per 0ea205f (actually ended up deleting this file altogether)...

khandelwal commented on forecast-admin/forecast/opportunities/fixtures/users.yaml:
```diff
@@ -0,0 +1,13 @@
+- fields:
```
The fixture above is in JSON. Would it make sense to stick to one format?

vdavez replied:
Fixed per 0ea205f (also deleted)...

khandelwal commented on forecast-admin/forecast/opportunities/fixtures/users.yaml:
```diff
@@ -0,0 +1,13 @@
+    is_superuser: true
+    email: ''
+    username: admin
+    password: pbkdf2_sha256$20000$SeBQtWDFH2HW$FPmflJ4hN0yk4uPS47gqMyqEt2WhSWFVxOqUFLRxuLE=
```
This gives me some pause. Is this necessary?

vdavez replied:
It isn't strictly necessary. I used it when I was doing local dev because I had to wipe my db enough. Happy to remove it.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+class Command(BaseCommand):
+    help = 'Load data from CSV'
+
+    default_filename = 'forecast/fixtures/fy16q1.csv'
```
The CSV file for the data, shouldn't be in the fixtures directory. Fixtures are a specific thing in Django. If anything, put it in a data/ directory.

vdavez replied:
Fixed per 39be4b9.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+            '-f', '--filename',
+            default=default_filename,
+            help='input filename (.csv, default {})'.format(
+                default_filename
```
I probably wouldn't do a default - so that we force people to think about what data they are uploading.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+        )
+
+
+class OpportunitiesLoader(object):
```
This doesn't really need to be a class.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+                except (ValueError, ValidationError) as e:
+                    if strict:
+                        print('error parsing {}'.format(row))
+                        raise
```
Raise an Exception here, with what's in the print as the message of the exception. That's probably better.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+                        print('error parsing {}'.format(row))
+                        raise
+                    else:
+                        skipped += 1
```
It doesn't look like you're doing anything with skipped?

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+
+class OpportunitiesLoader(object):
+    model = Opportunity
+    header_rows = 1
```
You don't need this - because there's no way to change this on the command line.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+            published=True
+        )
+        return opportunity
+    #     schedule = row[9]
```
Commented out code should just be deleted.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+    #         return fallback
+    #
+
+    @staticmethod
```
These static methods can just be functions. They don't need to belong to the class.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+
+    @staticmethod
+    def parse_date(s):
+        if not s or not re.match(r"\d{2}\/\d{2}\/\d{2,4}", s):
```
I'd use strptime to parse this date (or whatever the actual Python function is). There's no reason to use a regular expression.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+    #
+
+    @staticmethod
+    def parse_date(s):
```
This method needs a docstring.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+        return date(year, month, day)
+
+    @staticmethod
+    def parse_dollars(s):
```
Docstring missing here.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+                region=region
+            )
+            return obj
+        except ValueError:
```
Instead of trapping the exception here and returning False, I would just let the exception bubble up and deal with it in the calling code if you need to. Hiding exceptions like this can be tricky.

khandelwal commented on forecast-admin/forecast/opportunities/management/commands/load_opportunities.py:
```diff
@@ -0,0 +1,193 @@
+                default_filename
+            )
+        ),
+        make_option(
```
I know why this is here - but generally I like data loading functions to be idempotent. If you run this twice in append mode, you're doing to get data repeats right?

khandelwal commented on forecast-admin/forecast/opportunities/models.py:
```diff
@@ -12,9 +15,22 @@ def __str__(self):
 
 
-class Award(models.Model):
+class OSBUAdvisor(models.Model):
```
Add a docstring here to explain what an OSBUAdvisor is.

vdavez replied:
Done per 6ef0a3c

khandelwal commented:
It also looks like there are no tests for the new Python code. Is that correct, or did I miss them? We'll need some tests before merging.

khandelwal commented:
I'm done reviewing the Python code. I have no expertise to review the front-end pieces.

Commit:
Admin improvements

1. Moved the admin actions into the Opportunity ModelAdmin
2. Added test coverage for the OSBUAdvisor admin
3. Removed the `pass` for OSBUAdvisor

Commit:
Removed create_offices fixture

Initially, I imagined generating all of the possible permutations of
offices that could exist within GSA. Now, however, I think it's time
to deprecate it and let the load_opportunities command to generate
offices that actually have contracting activity. (i.e., OCSIT doesn't
have a R9 office that separately procures).

Commit:
moved fy16 data into a data folder

Commit:
added description of OSBUAdvisor

Commit:
removed load_command from branch

vdavez commented:
@khandelwal With the exception of the changes to the `load_opportunities` command, I have made changes here. In the interest of keeping things clean and moving, I  decided to separate the command and test into a separate branch, and deleted from this branch per ffb139f.

Commit:
merge conflict

vdavez commented on forecast-admin/forecast/opportunities/admin.py:
```diff
@@ -1,12 +1,26 @@
 @admin.register(Opportunity)
 class OpportunityAdmin(admin.ModelAdmin):
 
+    def make_published(self, request, queryset):
```
@khandelwal Any idea how to write unit tests for this? I've spent some time in the django docs and in stackoverflow but am coming up broke for actual answers.

khandelwal replied:
I'm sure there are complicated methods, a simple one would be: 

Create a function (outside of the class) that just takes in a query set, and marks everything in the queryset as published. That should be easy to test.  This method doesn't need the request, or self.

vdavez commented on forecast-admin/forecast/opportunities/models.py:
```diff
@@ -160,46 +221,49 @@ class Opportunity(models.Model):
-    estimated_fiscal_year_quarter = models.CharField(max_length=200,
-                                                     blank=True, null=True)
+    # TODO: going to make these choices...
+    estimated_fiscal_year = models.IntegerField(default=2016)
```
Related. When you do unit tests, do you test for things like model fields? I mean, I _could_ write a function that tests whether it fails to create an Opportunity if I used `2016a`, but I've been assuming that Django will handle that for me. :grimacing:

khandelwal commented on forecast-admin/forecast/opportunities/admin.py:
```diff
@@ -1,12 +1,26 @@
+        for obj in queryset:
+            obj.published = True
+            obj.save()
+    make_published.short_description = "Publish selected oppotunities"
```
Spelling: should be 'opportunities' - you're missing the 'r'

khandelwal commented:
I'm going to go-ahead and merge this, as I hate leaving long pull requests open for lengthy periods of time. Everything can be addressed in subsequent efforts.
